maven-core/src/main/java/org/apache/maven/internal/transformation/TransformedArtifact.java

```diff
@@ -67,36 +71,36 @@ public void setFile(File file) {
 
     @Override
     public File getFile() {
-        Path result = onChangeTransformer.get();
-        if (result == null) {
-            return null;
+        try {
+            Path result = onChangeTransformer.transform();
+            if (result == null) {
+                return null;
+            }
+            return result.toFile();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
         }
-        return result.toFile();
     }
 
     private static final int BUFFER_SIZE = 8192;
 
-    private static String sha1(Path path) {
-        try {
-            MessageDigest md = MessageDigest.getInstance("SHA-1");
-            try (InputStream fis = Files.newInputStream(path)) {
-                byte[] buffer = new byte[BUFFER_SIZE];
-                int read;
-                while ((read = fis.read(buffer)) != -1) {
-                    md.update(buffer, 0, read);
-                }
-            }
-            StringBuilder result = new StringBuilder();
-            for (byte b : md.digest()) {
-                result.append(String.format("%02x", b));
+    private static String sha1(Path path) throws Exception {
+        MessageDigest md = MessageDigest.getInstance("SHA-1");
+        try (InputStream fis = Files.newInputStream(path)) {
+            byte[] buffer = new byte[BUFFER_SIZE];
+            int read;
+            while ((read = fis.read(buffer)) != -1) {
+                md.update(buffer, 0, read);
             }
-            return result.toString();
-        } catch (Exception e) {
-            throw new RuntimeException(e);
         }
+        StringBuilder result = new StringBuilder();
+        for (byte b : md.digest()) {
+            result.append(String.format("%02x", b));
+        }
+        return result.toString();
     }
 
-    protected abstract void transform(Path src, Path dst);
+    protected abstract void transform(Path src, Path dst) throws Exception;
```

Comment:
Maybe an IOException, or maybe some redesign such that no exceptions are thrown. This doesn't need any possible exception

Reply:
Not sure what you mean.  If I don't throw a checked exception, we get back to the fact you don't like throwing unchecked exceptions. I'm really lost between your conflicting statements.  The full list of checked exceptions that can be checked by the underlying code is the following: `IOException, NoSuchAlgorithmException, XMLStreamException, ModelBuildingException, ComponentLookupException`.

Reply:
In general, do not throw java.lang.Exception. Use a more specific subclass that's appropriate to that point in the code. See Items 72 and 73 of Effective Java. 

But it might be possible to not throw any exceptions here, checked or runtime. It depends a little on what the method is expected to do. The trick would be to avoid reading files during the transformation and do that before and/or after the transform. Separate the transform from I/O.

Reply:
That forces the definition of lots of exceptions, just for the sake of avoiding `throws Exception`.  I think I don't really agree to this.
Anyway, I refactored the whole package to avoid exception chaining and everything is caught in a public specific exception.